nemo/collections/speechlm2/models/duplex_ear_tts.py

```diff
@@ -0,0 +1,1991 @@
+        audio = torch.zeros(1, 10 * self.target_sample_rate).float().to(self.device)
+        audio_len = torch.tensor([audio.size(-1)]).long()
+        audio, audio_len = self.pad_audio_to_factor(audio, audio_len, self.target_samples_per_frame)
+
+        with ensures_target_precision(self.audio_codec_run_dtype), torch.no_grad():
+            sil_codes, _ = self.audio_codec.encode(audio.unsqueeze(1), audio_len)  # [1, T, C]
+            sil_codes = sil_codes[0]  # [T, C]
+
+        # Convert each frame (C tokens) into a tuple
+        combos = [tuple(row.tolist()) for row in sil_codes]
+
+        # Count frequencies
+        counter = Counter(combos)
+
+        # Pick the most common combination
+        most_common_combo, freq = counter.most_common(1)[0]
+
+        # Return as tensor [C]
+        return torch.tensor(most_common_combo, device=self.device, dtype=torch.long)
+
+    def _load_embed_tokens(self, cfg) -> nn.Embedding:
+        """Load token embedding layer for RVQ-EAR-TTS."""
+        if self.language_model:
+            assert callable(self.language_model.get_input_embeddings)
+            embed_tokens: nn.Embedding = self.language_model.get_input_embeddings()
+        else:
+            embed_tokens_state_dict = torch.load(
+                cfg.pretrained_lm_embedding_path, map_location="cpu", weights_only=True
+            )
+
+            # Create token embedding layer
+            vocab_size, hidden_size = embed_tokens_state_dict["weight"].size()
+            embed_tokens = nn.Embedding(vocab_size, hidden_size, dtype=torch.bfloat16)
+            embed_tokens.load_state_dict(embed_tokens_state_dict)
+        return embed_tokens
+
+    def _load_tts_model(self, cfg) -> nn.Module:
+        """Load TTS model for RVQ-EAR-TTS."""
+        if self.cfg.get("pretrained_tts_model", None):
+            self.tts_model = RVQEARTTSModel.from_pretrained(
+                cfg.pretrained_tts_model, DictConfig(cfg.tts_config), strict=False
+            )
+        else:
+            # start the model from scratch
+            self.tts_model = RVQEARTTSModel(DictConfig(cfg.tts_config))
+
+        setup_audio_codec(self)
+
+    def _load_language_model(self, cfg):
+        """Load language model for RVQ-EAR-TTS."""
+        if cfg.pretrained_lm_name:
+            language_model = load_pretrained_hf(
+                self.cfg.pretrained_lm_name, pretrained_weights=True, trust_remote_code=True
+            ).eval()
+        else:
+            language_model = None
+        return language_model
+
+    def restore_from_pretrained_checkpoint(self, checkpoint_path):
+        """
+        Loads model weights a pretrained checkpoint file, supporting partial loading from .nemo and PyTorch formats.
+
+        Args:
+            checkpoint_path (str): Path to checkpoint file.
+
+        Returns:
+            None. The model is updated in-place.
+        """
+        if checkpoint_path is not None:
+            if '.nemo' in checkpoint_path:
```

Comment:
Why would `.nemo` be in checkpoint_path? speechlm2 doesn't save .nemo checkpoints, it saves in .safetensors

Reply:
It was to be compatible with the older codebase. Perhaps we can remove it now, given that everyone moved to the new speechlm2-based codebase.